File: ports/java/src/main/java/com/codinglombok/lombokalgoritma/sort/Sort.java
// LombokAlgoritma — Java Sort Module — Apache-2.0 — @codinglombok
package com.codinglombok.lombokalgoritma.sort;
import java.util.Arrays; import java.util.Comparator;
public final class Sort {
  private Sort(){}
  public static <T extends Comparable<T>> T[] timsort(T[] a){T[] r=Arrays.copyOf(a,a.length);Arrays.sort(r);return r;}
  public static <T> T[] timsort(T[] a,Comparator<T> c){T[] r=Arrays.copyOf(a,a.length);Arrays.sort(r,c);return r;}
  public static int[] mergesort(int[] arr){
    int[]a=arr.clone(),n=new int[]{a.length},tmp=new int[a.length];
    for(int w=1;w<n[0];w*=2){for(int lo=0;lo<n[0];lo+=2*w){int mid=Math.min(lo+w,n[0]),hi=Math.min(lo+2*w,n[0]),i=lo,j=mid,k=lo;
      while(i<mid&&j<hi){if(a[i]<=a[j])tmp[k++]=a[i++];else tmp[k++]=a[j++];}while(i<mid)tmp[k++]=a[i++];while(j<hi)tmp[k++]=a[j++];}
      System.arraycopy(tmp,0,a,0,n[0]);}return a;}
  public static int[] quicksort(int[] arr){int[]a=arr.clone();if(a.length>1)qs(a,0,a.length-1);return a;}
  private static void qs(int[]a,int lo,int hi){if(lo>=hi)return;if(hi-lo<16){is(a,lo,hi);return;}int p=part(a,lo,hi);qs(a,lo,p-1);qs(a,p+1,hi);}
  private static int part(int[]a,int lo,int hi){int mid=(lo+hi)/2;if(a[lo]>a[mid]){int t=a[lo];a[lo]=a[mid];a[mid]=t;}if(a[lo]>a[hi]){int t=a[lo];a[lo]=a[hi];a[hi]=t;}if(a[mid]>a[hi]){int t=a[mid];a[mid]=a[hi];a[hi]=t;}int p=a[hi],i=lo;for(int j=lo;j<hi;j++)if(a[j]<=p){int t=a[i];a[i]=a[j];a[j]=t;i++;}int t=a[i];a[i]=a[hi];a[hi]=t;return i;}
  private static void is(int[]a,int lo,int hi){for(int i=lo+1;i<=hi;i++){int k=a[i],j=i-1;while(j>=lo&&a[j]>k){a[j+1]=a[j];j--;}a[j+1]=k;}}
  public static int[] heapsort(int[] arr){int[]a=arr.clone();int n=a.length;for(int i=n/2-1;i>=0;i--)sft(a,i,n);for(int e=n-1;e>0;e--){int t=a[0];a[0]=a[e];a[e]=t;sft(a,0,e);}return a;}
  private static void sft(int[]a,int r,int e){while(true){int lg=r,l=2*r+1,ri=2*r+2;if(l<e&&a[l]>a[lg])lg=l;if(ri<e&&a[ri]>a[lg])lg=ri;if(lg==r)break;int t=a[r];a[r]=a[lg];a[lg]=t;r=lg;}}
  public static int[] countingSort(int[] arr){if(arr.length==0)return new int[0];int max=0;for(int v:arr)if(v>max)max=v;int[]cnt=new int[max+1];for(int v:arr)cnt[v]++;int[]out=new int[arr.length],k=new int[]{0};for(int i=0;i<=max;i++)for(int j=0;j<cnt[i];j++)out[k[0]++]=i;return out;}
}

File: ports/kotlin/src/commonMain/kotlin/com/codinglombok/lombokalgoritma/Sort.kt
// LombokAlgoritma — Kotlin Sort Module
// Apache-2.0 — @codinglombok
package com.codinglombok.lombokalgoritma

/**
 * Timsort — stable, adaptive. O(n log n).
 * Delegates to Kotlin stdlib's sortedWith which uses TimSort.
 */
fun <T : Comparable<T>> timsort(arr: List<T>): List<T> = arr.sorted()

fun <T> timsortWith(arr: List<T>, comparator: Comparator<T>): List<T> = arr.sortedWith(comparator)

/** Mergesort — bottom-up iterative, O(n log n), stable. */
fun <T : Comparable<T>> mergesort(arr: List<T>): List<T> {
    val a = arr.toMutableList()
    val n = a.size
    if (n <= 1) return a
    val tmp = MutableList(n) { a[it] }
    var width = 1
    while (width < n) {
        var lo = 0
        while (lo < n) {
            val mid = minOf(lo + width, n)
            val hi = minOf(lo + 2 * width, n)
            var i = lo; var j = mid; var k = lo
            while (i < mid && j < hi) {
                if (a[i] <= a[j]) tmp[k++] = a[i++] else tmp[k++] = a[j++]
            }
            while (i < mid) tmp[k++] = a[i++]
            while (j < hi) tmp[k++] = a[j++]
            lo += 2 * width
        }
        for (idx in 0 until n) a[idx] = tmp[idx]
        width *= 2
    }
    return a
}

/** Quicksort — O(n log n) average. Not stable. */
fun <T : Comparable<T>> quicksort(arr: List<T>): List<T> {
    val a = arr.toMutableList()
    if (a.size <= 1) return a
    quicksortInner(a, 0, a.size - 1)
    return a
}

private fun <T : Comparable<T>> quicksortInner(a: MutableList<T>, lo: Int, hi: Int) {
    if (lo >= hi) return
    if (hi - lo < 16) { insertionSort(a, lo, hi); return }
    val p = partitionKt(a, lo, hi)
    quicksortInner(a, lo, p - 1)
    quicksortInner(a, p + 1, hi)
}

private fun <T : Comparable<T>> partitionKt(a: MutableList<T>, lo: Int, hi: Int): Int {
    val mid = (lo + hi) / 2
    if (a[lo] > a[mid]) { val t = a[lo]; a[lo] = a[mid]; a[mid] = t }
    if (a[lo] > a[hi])  { val t = a[lo]; a[lo] = a[hi];  a[hi] = t  }
    if (a[mid] > a[hi]) { val t = a[mid]; a[mid] = a[hi]; a[hi] = t }
    val t = a[mid]; a[mid] = a[hi]; a[hi] = t
    val pivot = a[hi]; var i = lo
    for (j in lo until hi) if (a[j] <= pivot) { val u = a[i]; a[i] = a[j]; a[j] = u; i++ }
    val u = a[i]; a[i] = a[hi]; a[hi] = u
    return i
}

private fun <T : Comparable<T>> insertionSort(a: MutableList<T>, lo: Int, hi: Int) {
    for (i in lo + 1..hi) {
        val key = a[i]; var j = i - 1
        while (j >= lo && a[j] > key) { a[j + 1] = a[j]; j-- }
        a[j + 1] = key
    }
}

/** Heapsort — in-place, O(n log n). Not stable. */
fun <T : Comparable<T>> heapsort(arr: List<T>): List<T> {
    val a = arr.toMutableList(); val n = a.size
    if (n <= 1) return a
    for (i in n / 2 - 1 downTo 0) sift(a, i, n)
    for (end in n - 1 downTo 1) { val t = a[0]; a[0] = a[end]; a[end] = t; sift(a, 0, end) }
    return a
}

private fun <T : Comparable<T>> sift(a: MutableList<T>, root: Int, end: Int) {
    var r = root
    while (true) {
        var largest = r; val l = 2*r+1; val ri = 2*r+2
        if (l < end && a[l] > a[largest]) largest = l
        if (ri < end && a[ri] > a[largest]) largest = ri
        if (largest == r) break
        val t = a[r]; a[r] = a[largest]; a[largest] = t; r = largest
    }
}
